### PR TITLE
Leave one out cross validation

### PR DESCRIPTION
**Pull request by mortonjt**
There are a few things that are added to this PR namely

- percent_explained : calculates the percentage of the total variance explained for each balance axis (think about the analogue for principal components).

- leave one out cross validation : calculates model error on training dataset and prediction error on testing dataset

- leave one variable out cross validation : calculates R^2 and model error when each variable is left out of the model.  This gives us an idea about the variable importance for each variable.  But more work concerning effect size of each variable will need to be done in the future

In addition, the OLS procedure was slightly optimized to reduce the amount of formula -> matrix conversions that needed to be computed.  Will need to propagate this to the remaining methods.  Thanks @Josh-Lefler for discovering this!

Remember, this is the last major PR that needs to be merged in.  Reviews will be greatly appreciated.  Thanks!

### CODE REVIEW

File: gneiss/regression/_ols.py [hunk 1]
```diff
@@ -7,13 +7,28 @@
...
 import statsmodels.formula.api as smf
 from statsmodels.iolib.summary2 import Summary
+from statsmodels.sandbox.tools.cross_val import LeaveOneOut
+from patsy import dmatrix
+
+
+def _fit_ols(y, x, **kwargs):
+    """ Perform the basic ols regression."""
+    exog_data = x
+    submodels = []
```
Comment (josenavas) [inline gneiss/regression/_ols.py]: Minor, and it depends on how big `len(y.columns)` is but this can be written as:
```python
submodels = [smf.OLS(endog=y[b], exog=x, **kwargs) for b in y.columns]
```

Appending to a list may take more than expected if the list is really long.

File: gneiss/regression/_ols.py [hunk 2]
```diff
@@ -298,3 +318,144 @@ def r2(self):
 
         sst = sse + ssr
         return 1 - sse / sst
+
+    @property
+    def mse(self):
```
Comment (josenavas) [inline gneiss/regression/_ols.py]: Is this property described somewhere (i.e. doc)?

File: gneiss/regression/_ols.py [hunk 3]
```diff
@@ -298,3 +318,144 @@ def r2(self):
...
+           pred_err : np.array, float
+               Prediction mean sum of squares error for each iteration of
+               the cross validation.
+
+        See Also
+        --------
+        fit
+        """
+        params = {'fit_regularized': False}
+        for key in params:
```
Comment (josenavas) [inline gneiss/regression/_ols.py]: Are you expecting to grow `params` in the future? As it is, this `for` loop only executes one operation so I don't see the point of having it.

File: gneiss/regression/_ols.py [hunk 4]
```diff
@@ -298,3 +318,144 @@ def r2(self):
...
+        nobs = self.balances.shape[0]  # number of observations (i.e. samples)
+        cv_iter = LeaveOneOut(nobs)
+        endog = self.balances
+        exog_names = self.results[0].model.exog_names
+        exog = pd.DataFrame(self.results[0].model.exog,
+                            index=self.balances.index,
+                            columns=exog_names)
+        results = pd.DataFrame(index=self.balances.index,
+                               columns=['mse', 'pred_err'])
+        i = 0
```
Comment (josenavas) [inline gneiss/regression/_ols.py]: I think there are 2 more pythonic way of doing this is, I'm unsure which one is the best for your case. Since `i` is just an index, this should work:
```python
for sample_id, (inidx, outidx) in zip(self.balances.index, cv_iter):
   etc...
```

However, If you still need `i` for something else that I may be missing, then this is the pythonic way of doing the index counting:

```python
for i, (inidx, outidx) in enumerate(cv_iter):
    ....
```
And remove the `i+=1` from the end of the for loop.

File: gneiss/regression/_ols.py [hunk 5]
```diff
@@ -298,3 +318,144 @@ def r2(self):
...
+    def lovo(self, **kwargs):
+        """ Leave one variable out cross-validation.
+
+        Calculates summary statistics for each iteraction of leave one variable
+        out cross-validation, specially `r2` and `mse` on entire model.
+        This technique is particularly useful for feature selection.
+
+        Parameters
+        ----------
+        regularized : bool
```
Comment (josenavas) [inline gneiss/regression/_ols.py]: `regularized` is not named parameter of this function

File: gneiss/regression/_ols.py [hunk 6]
```diff
@@ -298,3 +318,144 @@ def r2(self):
...
+        pd.DataFrame
+           Rsquared : np.array, flot
+               Coefficient of determination for each variable left out.
+           mse : np.array, float
+               Mean sum of squares error for each iteration of
+               the cross validation.
+        """
+
+        params = {'regularized': False}
+        for key in params:
```
Comment (josenavas) [inline gneiss/regression/_ols.py]: Same comment as above regarding the for loop.

File: gneiss/regression/_ols.py [hunk 7]
```diff
@@ -298,3 +318,144 @@ def r2(self):
...
+        endog = self.balances
+        exog_names = self.results[0].model.exog_names
+        exog = pd.DataFrame(self.results[0].model.exog,
+                            index=self.balances.index,
+                            columns=exog_names)
+        cv_iter = LeaveOneOut(len(exog_names))
+        results = pd.DataFrame(index=exog_names,
+                               columns=['mse', 'Rsquared'])
+        i = 0
+        for inidx, outidx in cv_iter:
```
Comment (josenavas) [inline gneiss/regression/_ols.py]: Same comment as above regarding the usage of the `i` variable.

File: gneiss/regression/tests/test_ols.py [hunk 8]
```diff
@@ -226,6 +240,113 @@ def test_summary_head(self):
...
+        exp_coefs = pd.DataFrame({
+            'y0': [-2.074816e+09, -2.039271e+08, -1.843143e+08,
+                   2.261170e+09, -7.925699e+06],
+            'y1': [-2.074816e+10, -2.039271e+09, -1.843143e+09,
+                   2.261170e+10, -7.925699e+07]
+            }, index=['Intercept', 'x1', 'x2', 'x3', 'x4']).T
+        # The L1 regularization either has numerical issues
+        # or is random ...
+        res_coefs = res.coefficients()
+        pdt.assert_index_equal(exp_coefs.index, res_coefs.index)
```
Comment (josenavas) [inline gneiss/regression/tests/test_ols.py]: Minor and I'm unsure if it applies in the pandas testing functions: in general we use `assert*(observed, expected)`, but I'm unsure if the pandas error messages are expecting this convention.

File: gneiss/regression/tests/test_ols.py [hunk 9]
```diff
@@ -226,6 +240,113 @@ def test_summary_head(self):
...
+        res = ols(formula="x1 + x2 + x3 + x4",
+                  table=self.y, metadata=self.x, tree=self.t2)
+        res.fit(regularized=True)
+        exp_coefs = pd.DataFrame({
+            'y0': [-2.074816e+09, -2.039271e+08, -1.843143e+08,
+                   2.261170e+09, -7.925699e+06],
+            'y1': [-2.074816e+10, -2.039271e+09, -1.843143e+09,
+                   2.261170e+10, -7.925699e+07]
+            }, index=['Intercept', 'x1', 'x2', 'x3', 'x4']).T
+        # The L1 regularization either has numerical issues
```
Comment (josenavas) [inline gneiss/regression/tests/test_ols.py]: From my ignorant view, if it is random, can you set a seed? Otherwise, can you test some of the values up to a certain decimal position?

File: gneiss/regression/_ols.py (hunk 1 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: I didn't know that.  Thanks!

File: gneiss/regression/_ols.py (hunk 2 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: I'll add it in.  It is a measure of Mean of Sum of Squares Error, which is used for calculating R^2 and the cross validation statistics.

File: gneiss/regression/_ols.py (hunk 3 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: The main reason for having the kwargs argument was to pass these into the `fit` function, which have a ton of parameters in them.  One advantage of doing this is if the `fit` function in statsmodels changes, we aren't forced to change our APIs to fit their changes.

File: gneiss/regression/_ols.py (hunk 4 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: Will use enumerate instead.  Thanks!

File: gneiss/regression/_ols.py (hunk 5 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: I'll move this up to the params section.  We do actually need this

File: gneiss/regression/tests/test_ols.py (hunk 9 above)
Comment (mortonjt) [inline gneiss/regression/tests/test_ols.py]: So this is actually depending on https://github.com/statsmodels/statsmodels/issues/3393

Not sure what is going on there - but maybe we shouldn't include any regularization fitting here until that issue gets resolved.  I'm going to remove those fits in the next commit.

File: gneiss/regression/tests/test_ols.py (hunk 8 above)
Comment (mortonjt) [inline gneiss/regression/tests/test_ols.py]: Going to remove this test since we won't be needing it.

File: gneiss/regression/_ols.py (hunk 5 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: Scratch that.  I'll delete it.

File: gneiss/regression/_ols.py [hunk 10]
```diff
@@ -203,11 +210,24 @@ def __init__(self, *args, **kwargs):
...
+            when performing the fit. (default = False)
+        **kwargs : dict
+           Keyword arguments used to tune the parameter estimation.
+
+        """
         for s in self.submodels:
             # assumes that the underlying submodels have implemented `fit`.
-            m = s.fit(**kwargs)
+            if regularized:
+                m = s.fit(**kwargs)
```
Comment (mortonjt) [inline gneiss/regression/_ols.py]: This is backwards.  This should be `fit_regularized`,  not `fit`.  I'm removing the `fit_regularized` functionality for the time being to resolve this.  But this actually messed up the `test_mse` test.  I have fixed that in the upcoming commit

**Comment by josenavas** [comment]
Once tests pass

**Comment by mortonjt** [comment]
@sjanssen2 @qiyunzhu could you review / merge?  Thanks!

File: gneiss/regression/_ols.py [hunk 11]
```diff
@@ -7,13 +7,23 @@
...
 from statsmodels.iolib.summary2 import Summary
+from statsmodels.sandbox.tools.cross_val import LeaveOneOut
+from patsy import dmatrix
+
+
+def _fit_ols(y, x, **kwargs):
+    """ Perform the basic ols regression."""
+    # mixed effects code is obtained here:
+    # http://stackoverflow.com/a/22439820/1167475
+    submodels = [smf.OLS(endog=y[b], exog=x, **kwargs) for b in y.columns]
```
Comment (qiyunzhu) [inline gneiss/regression/_ols.py]: Maybe these two lines can be merged into one? (Not sure if what I proposed is Pythonic, if not sorry.)

File: gneiss/regression/_ols.py [hunk 12]
```diff
@@ -203,12 +205,20 @@ def __init__(self, *args, **kwargs):
         """
         super().__init__(*args, **kwargs)
 
-    def fit(self, **kwargs):
-        """ Fit the model """
-        for s in self.submodels:
-            # assumes that the underlying submodels have implemented `fit`.
-            m = s.fit(**kwargs)
-            self.results.append(m)
+    def fit(self, regularized=False, **kwargs):
```
Comment (qiyunzhu) [inline gneiss/regression/_ols.py]: So this rewritting is basically an optimization of the old function. Is that right?

File: gneiss/regression/_ols.py [hunk 13]
```diff
@@ -295,6 +303,122 @@ def r2(self):
...
+        exog_names = self.results[0].model.exog_names
+        exog = pd.DataFrame(self.results[0].model.exog,
+                            index=self.balances.index,
+                            columns=exog_names)
+        cv_iter = LeaveOneOut(len(exog_names))
+        results = pd.DataFrame(index=exog_names,
+                               columns=['mse', 'Rsquared'])
+        for i, (inidx, outidx) in enumerate(cv_iter):
+            feature_id = exog_names[i]
+            res_i = _fit_ols(endog, exog.loc[:, inidx], **kwargs)
```
Comment (qiyunzhu) [inline gneiss/regression/_ols.py]: This part is bit awkward to me. `res_i` is defined twice with different data types. Is this Pythonic?

File: gneiss/regression/_ols.py [hunk 14]
```diff
@@ -295,6 +303,122 @@ def r2(self):
...
+            # sum of squares regression.
+            ssr = sum([r.ess for r in res_i])
+            # sum of squares error.
+            sse = sum([r.ssr for r in res_i])
+            # calculate the overall coefficient of determination (i.e. R2)
+            sst = sse + ssr
+            results.loc[feature_id, 'Rsquared'] = 1 - sse / sst
+            # degrees of freedom for residuals
+            dfe = res_i[0].df_resid
+            results.loc[feature_id, 'mse'] = sse / dfe
```
Comment (qiyunzhu) [inline gneiss/regression/_ols.py]: Can this `dfe` be zero?

File: gneiss/regression/_ols.py [hunk 15]
```diff
@@ -295,6 +303,122 @@ def r2(self):
...
+        return results
+
+    def percent_explained(self):
+        """ Proportion explained by each principal balance."""
+        # Using sum of squares error calculation (df=1)
+        # instead of population variance (df=0).
+        axis_vars = np.var(self.balances, ddof=1, axis=0)
+        return axis_vars / axis_vars.sum()
```
Comment (qiyunzhu) [inline gneiss/regression/_ols.py]: Can this `axis_vars.sum()` be zero?

File: gneiss/regression/_ols.py (hunk 14 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: If that ever happens, would indicate overfitting to the extreme (i.e. more parameters than data points).  I think it is safe to say that this will never happen.

File: gneiss/regression/_ols.py (hunk 15 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: That would mean that the variance is zero.  For practical reasons, that will not happen.

File: gneiss/regression/_ols.py (hunk 12 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: yup!

File: gneiss/regression/_ols.py (hunk 11 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: meh sure.

File: gneiss/regression/_ols.py (hunk 13 above)
Comment (mortonjt) [inline gneiss/regression/_ols.py]: ok, I'll rename it

Comment (mortonjt) [inline gneiss/regression/_ols.py]: Basically, all this is doing is

1. Retrieving the model specifications for the linear regression (`model_i`)
2. Actually fitting the model with a numerical optimization (`res_i`)

Comment (qiyunzhu) [inline gneiss/regression/_ols.py]: Sounds good!

File: gneiss/regression/_ols.py (hunk 14 above)
Comment (qiyunzhu) [inline gneiss/regression/_ols.py]: Got it!

File: gneiss/regression/_ols.py (hunk 15 above)
Comment (qiyunzhu) [inline gneiss/regression/_ols.py]: I see.
